src/main/msp/msp_serial.c

```diff
@@ -131,6 +133,36 @@ static void mspSerialResponse(mspPort_t *msp, mspPacket_t *reply)
     serialEndWrite(msp->port);
 }
 
+
+static mspPacket_t *mspCreateMessage(void)
+{
+    static uint8_t outBuf[MSP_PORT_OUTBUF_SIZE];
+    static mspPacket_t message = {
+        .buf = {
+            .ptr = outBuf,
+            .end = ARRAYEND(outBuf),
+        },
+        .cmd = -1,
+        .result = 0,
+    };
+
+    return &message;
+}
+
+static mspPacket_t *mspCreateOutgoing(void)
+{
+    mspPacket_t *reply = mspCreateMessage();
+    reply->direction = OUTGOING;
+    return reply;
+}
+
+static mspPacket_t *mspCreateIncoming(void)
+{
+    mspPacket_t *reply = mspCreateMessage();
+    reply->direction = INCOMING;
+    return reply;
+}
+
 STATIC_UNIT_TESTED void mspSerialProcessReceivedCommand(mspPort_t *msp)
```

Comment:
msp port is unidirectional now, but this function hides it. If unexpected packet is received(for example due to shorted RX-TX resulting in echo), this function will handle it to wrong destination.

It would be better to decide direction unconditionally according to MSP port type.